# 谜题 44：切掉类  

## 背景

 NoClassDefFoundError。

## 代码

请考虑下面的两个类：  

```java
public class Strange1 {
    public static void main(String[] args) {
        try {
            Missing m = new Missing();
        } catch (java.lang.NoClassDefFoundError ex) {
            System.out.println("Got it!");
        }
    }
}

public class Strange2 {
    public static void main(String[] args) {
        Missing m;
        try {
            m = new Missing();
        } catch (java.lang.NoClassDefFoundError ex) {
            System.out.println("Got it!");
        }
    }
}

//Strange1 和 Strange2 都用到了下面这个类：
class Missing {
    Missing() { 
    }
}
```

## 陷阱

如果你编译所有这三个类，然后在运行 Strange1 和 Strange2 之前删除Missing.class 文件，你就会发现这两个程序的行为有所不同。

程序 Strange1 只在其 try 语句块中提及 Missing 类型，因此你可能会认为它捕获 NoClassDefFoundError 异常，并打印 Got it！另一方面，程序 Strange2 在try 语句块之外声明了一个 Missing 类型的变量，因此你可能会认为所产生的NoClassDefFoundError 异常不会被捕获。

如果你试着运行这些程序，就会看到它们的行为正好相反：Strange1 抛出了未被捕获的 NoClassDefFoundError 异常，而 Strange2 却打印出了 Got it！怎样才能解释这些奇怪的行为呢？  

## 解惑

如果你去查看 Java 规范以找出应该抛出 NoClassDefFoundError 异常的地方，那么你不会得到很多的指导信息。该规范描述道，这个错误可以“在（直接或间接）使用某个类的程序中的任何地方”抛出。

当 VM 调用 Strange1 和Strange2 的 main 方法时，这些程序都间接使用了 Missing 类，因此，它们都在其权利范围内于这一点上抛出了该错误。  

要查明为什么会是这样，我们需要研究一下由编译器生成的这些程序的字节码。  略...........

因为 Missing.class 文件已经被删除了，所以校验器不能加载它，因而抛出了一个 NoClassDefFoundError 异常。请注意，这个异常
是在校验期间、在类被初始化之前，并且在 main 方法开始执行之前很早就抛出的。这就解释了为什么没有打印出任何关于这个未被捕获异常的跟踪栈信息。  

## 解决办法或规则

**教训**：要想编写一个能够探测出某个类是否丢失的程序，请使用反射来引用类而不要使用通常的语言结构。  

```java
public class Strange {
    public static void main(String[] args) throws Exception{
        try {
            Object m = Class.forName("Missing").newInstance();
        } catch (ClassNotFoundException ex) {
            System.err.println("Got it!");
        }
    }
}
```

总之，不要对捕获 NoClassDefFoundError 形成依赖。语言规范非常仔细地描述了类初始化是在何时发生的，但是类被加载的时机却显得更加不可预测。

更一般地讲，**捕获 Error 及其子类型几乎是完全不恰当的，这些异常是为那些不能被恢复的错误而保留的**。  